src/Microsoft.Diagnostics.DebugServices.Implementation/SymbolService.cs

```diff
@@ -0,0 +1,249 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.FileFormats;
+using Microsoft.SymbolStore;
+using Microsoft.SymbolStore.KeyGenerators;
+using SOS;
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Reflection.PortableExecutable;
+using System.Text;
+
+namespace Microsoft.Diagnostics.DebugServices.Implementation
+{
+    /// <summary>
+    /// Symbol services to configure symbol servers, caches and directory search
+    /// </summary>
+    public class SymbolService : ISymbolService
+    {
+        private readonly IHost _host;
+
+        public SymbolService(IHost host)
+        {
+            _host = host;
+        }
+
+        /// <summary>
+        /// Invoked when anything changes in the symbol service (adding servers, caches, or directories, clearing store, etc.)
+        /// </summary>
+        public event ISymbolService.ChangeEventHandler OnChangeEvent;
```

Comment:
Can we implement an interface instead of using events in general and use a clean method? The callbacks seem to only use the state of the object they are cleaning. events in my experience are very hard to diagnose, and right now they are rooting a large amount of locals and could leak as long as a target or symbol service is kept alive somehow.

Reply:
I'll do some better in Part 3.  I don't like events either.  I've been thinking about a simple IServiceEvent interface with a Register method that takes a call back, returns a IDisposable (used to unregister) and keeps the registration on a weak list so it will get unregistered if the calling object gets collected.